bootstrap/controllers/certificates_controller.go

```diff
@@ -0,0 +1,219 @@
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	bsutil "sigs.k8s.io/cluster-api/bootstrap/util"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// CertificatesReconciler reconciles a Machine object
+type CertificatesReconciler struct {
+	client.Client
+	Log      logr.Logger
+	Scheme   *runtime.Scheme
+	recorder record.EventRecorder
+
+	K8sdDialTimeout time.Duration
+
+	managementCluster ck8s.ManagementCluster
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *CertificatesReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if _, err := ctrl.NewControllerManagedBy(mgr).For(&clusterv1.Machine{}).Build(r); err != nil {
+		return err
+	}
+
+	r.Scheme = mgr.GetScheme()
+	r.recorder = mgr.GetEventRecorderFor("ck8s-certificates-controller")
+
+	if r.managementCluster == nil {
+		r.managementCluster = &ck8s.Management{
+			Client:          r.Client,
+			K8sdDialTimeout: r.K8sdDialTimeout,
+		}
+	}
+	return nil
+}
+
+type CertificatesScope struct {
+	Cluster  *clusterv1.Cluster
+	Config   *bootstrapv1.CK8sConfig
+	Log      logr.Logger
+	Machine  *clusterv1.Machine
+	Patcher  *patch.Helper
+	Workload *ck8s.Workload
+}
+
+// +kubebuilder:rbac:groups=bootstrap.cluster.x-k8s.io,resources=ck8sconfigs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=bootstrap.cluster.x-k8s.io,resources=ck8sconfigs/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;clusters/status;machines;machines/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups=exp.cluster.x-k8s.io,resources=machinepools;machinepools/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=secrets;events;configmaps,verbs=get;list;watch;create;update;patch;delete
+func (r *CertificatesReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("namespace", req.Namespace, "machine", req.Name)
+
+	m := &clusterv1.Machine{}
+	if err := r.Get(ctx, req.NamespacedName, m); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	if !m.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Machine is being deleted, return early.
+		return ctrl.Result{}, nil
+	}
+
+	mAnnotations := m.GetAnnotations()
+
+	var refreshCertificates, updateExpiryDate bool
+	_, refreshCertificates = mAnnotations[bootstrapv1.CertificatesRefreshAnnotation]
+	_, updateExpiryDate = mAnnotations[bootstrapv1.MachineCertificatesExpiryDateAnnotation]
+	if !refreshCertificates && updateExpiryDate {
```

Comment:
I think you meant for both to be negative?
```suggestion
	if !refreshCertificates && !updateExpiryDate {
```

Reply:
Please see this comment: https://github.com/canonical/cluster-api-k8s/pull/58#discussion_r1777384713